### DISCUSSION TITLE
Pi Zero 2W cannot see MCU

### DISCUSSION THREAD

**Discussion by chriskaye** [mainsail-crew/mainsail#2032] [Help]
Hi everyone,
I’m looking for a little help to point me in the right direction for an issue I have with an rPi Zero 2W and connecting the MCU with USB.

About a week ago, I installed Mainsail for the first time and have a working setup with Mainsail on a Raspberry Pi 3B+ connected by USB to a BigTreeTech SKR mini v1.1 board and it works great! =)

I wanted to free up the Pi 3B+ for other projects, and thought it would be a trivial matter to move my config files from the Pi 3B+ to a fresh Mainsail install on the Pi Zero 2W. How wrong I was!

Mainsail and the printer.cfg etc has been setup on the Pi Zero 2W, but when I connect the MCU via USB to a USB OTG-hub on the Pi Zero 2W, nothing happens.
The Pi Zero 2W doesn’t show any connection attempts at all (using dmesg and lsusb).

The OTG-hub is working (and set to host mode correctly) and can see USB thumb drives and various webcams that I connect (these show correctly in dmesg and lsusb) so the OTG-Hub is working without issue, but I don’t get why there are no registration events when connecting the working MCU to the OTG-Hub… 
I even made a micro USB male to USB type A cable pulling the OTG pin to ground and bypassing the OTG-Hub, and still nothing.

Every other write up I’ve seen about installing Mainsail, Klipper or Octopi on a Zero 2W shows it all should “just work” with minimal fuss.
Please, any words of wisdom would be appreciated!

Chris.

**Comment by meteyou** [comment]
I can only echo the other people. Whether you use a rpi3b or a rpi zero 2w makes no difference. You could even use the SD card directly from the rpi3b in the zero 2w without reflashing it.

If you can't see the board via dmesg or lsusb, you have a "physical" problem with the pi zero and the board. possibly a defective USB cable.

Are you sure it is an SKR mini v1.1 and not an SKR mini e3? Have you reflashed the board?

**Comment by chriskaye** [comment]
Hi, thanks for the reply.
It’s not a physical issue, the usb cable is the same as used on the rpi3b which works with the MCU just fine.
The OTG-hub works with other devices when connected to the Zero 2W.

The board is definitely a BTT SKR mini v1.1 (this board has sockets so you can swap stepper drivers).

When I flashed the firmware on the MCU, I didn’t change the boot loader, I’m thinking maybe this might help as on the rpi3b the MCU appears as 2 devices when connected, one is serial for the Klipper FW (I assume) and another device which I guess is the SD card that the flashed firmware file is located/local storage.
Maybe this is causing an issue with the reduced implementation of USB (thanks to the inclusion of the OTG stuff which is unnecessary for my usage scenario).
The only other alternatives would be uart serial over the GPIO pins or switching to an old 8bit board I have and ditching the BTT board…

I just plugged an old Melli board into the OTG-hub and the Zero 2W has picked it up straightaway in lsusb using the same cables - something funky is going on!

**Comment by meteyou** [comment]
> The board is definitely a BTT SKR mini v1.1 (this board has sockets so you can swap stepper drivers).

> When I flashed the firmware on the MCU, I didn’t change the boot loader, I’m thinking maybe this might help as on the rpi3b the MCU appears as 2 devices when connected, one is serial for the Klipper FW (I assume) and another device which I guess is the SD card that the flashed firmware file is located/local storage.

this question was only, because SKR mini e3 boards needs to set a pin to activate the usb port, but if you dont have a e3 board, you dont need it.

> I just plugged an old Melli board into the OTG-hub and the Zero 2W has picked it up straightaway in lsusb using the same cables - something funky is going on!

is your board connected to power right now? did you set the USB 5V jumper, if you dont have connected a power supply to it? do you see, if the boards "boot"?

**Reply by chriskaye** [reply]
I’m just running through creating a firmware for the old Melzi board using the Zero 2W.
It’s currently connected via USB power only on the OTG-Hub (yep jumper is set for this).

have tried with external power and without when using the BTT mini v1.1 board, still wasn’t detected as a device on the Zero’s usb bus…

**Comment by meteyou** [comment]
the firmware doesnt matter for lsusb or dmesg. so can you verify per LEDs or something similar, if the board is "booted"?

**Comment by chriskaye** [comment]
No, I don’t think there is any visible sign of the firmware booting (or has successfully booted).
The Mini v1 board just has a red LED that is permanently on when it has power.
The only thing I can think that would work is firing up rpi3b, connecting to MCU (to confirm firmware is booted) then unplug/re-plug into the rPi Zero 2W.
When I’ve tried this, it still doesn’t show in lsusb on the Zero 2, but plugging directly into the rPi3b it will show immediately (even without a power cycle).

**Reply by meteyou** [reply]
you dont need the firmware itself to boot/start. just the board should be connected with power and the MCU is "alive". then you should see it with lsusb or dmesg.

**Reply by chriskaye** [reply]
Yeah, I get what you are saying now, but I am unsure how to tell, maybe uart to the stm32 on the Mini v1.1 board to see if it is running when connected to the rPi Zero 2W (in case something in the OTG-hub is holding it in a reset state or preventing it from booting, maybe).

I just flashed the very old Melzi board and it has connected successfully as an MCU in a fresh install of Mainsail on the Zero 2W using the same OTG-Hub and USB cable. 
I can’t explain why the Zero 2W doesn’t want to talk to the Mini v1.1 board, as like you, I thought the rpi3b and Zero 2W were very close in terms of hardware (just less “extras” on the Zero 2W board…) - very strange!

**Reply by chriskaye** [reply]
OK, I have an update.
I think there is an issue with the bootloader on the SKR Mini v1.1
I reflashed it to an old backup of Marlin I had, and the device appears as a working COM port when connected to Windows.
When I reflash to Klipper, I can see a COM port connected to Windows and when I connect to it using putty, I just see jibberish in the console, as if my Baud, parity and other serial settings are configured incorrectly.
I tried many combinations (9600, 8, None, 1 all the way upto 500000 baud) but nothing worked.
For the old Marlin backup, I could connect using 115200, 8, N, 1 and issue G codes using the serial console.

The Pi Zero2W still doesn't recognise the BTT mini is connected to the USB port (no messages in dmesg).
My Pi3B and Windows do recognise the BTT mini and create a tty / COM port respectively.
I know the device is not registering correctly in Windows, as it would usually allow me to transfer new firmwares over USB to the onboard SD Card (which should appear in the File Explorer, but this no longer happens).
Looks like the device ID codes may not being sent from the BTT mini to the host, hence thinking of finding and reflashing a stock bootloader (if I can find one anywhere!).

This problem is not due to Pi, Klipper or Mainsail and either a freak occurrence or something dumb I did a while ago. 🙈 
If I get to the bottom of the issue, I'll provide an update for any other unfortunate souls in this situation.

@meteyou thank you for your assistance and maintaining Mainsail, I love it and think it is awesome.
